### ISSUE TITLE
ERROR: java.lang.NoClassDefFoundError: jakarta/servlet/http/HttpSessionContext with Jetty 11

### ISSUE THREAD

**Issue by ilapavuluri** [eclipse/jetty.project#11673]
**Jetty version(s)**
11.0.20

**Jetty Environment**
Jetty 11.0.20

**Java version/vendor** `(use: java -version)`
Java11
**OS type/version**
Windows
**Description**
I am upgrading my Java application to Jakarta (due to wildfly upgrade), I uses Jetty as well hence I am upgrading my Jetty from 9.x to 11.0.20, for now I dont want to upgrade to Jetty 12 since it requires to run on Java 17 and above.

I am getting below error:
java.lang.NoClassDefFoundError: jakarta/servlet/http/HttpSessionContext
	at deployment.c2oear-snapshot.ear/org.eclipse.jetty.servlet.ServletContextHandler.newSessionHandler(ServletContextHandler.java:343) ~[jetty-servlet-11.0.0.jar:11.0.0]
	at deployment.c2oear-snapshot.ear/org.eclipse.jetty.servlet.ServletContextHandler.getSessionHandler(ServletContextHandler.java:436) ~[jetty-servlet-11.0.0.jar:11.0.0]
	at deployment.c2oear-snapshot.ear/org.eclipse.jetty.servlet.ServletContextHandler.relinkHandlers(ServletContextHandler.java:262) ~[jetty-servlet-11.0.0.jar:11.0.0]
	at deployment.c2oear-snapshot.ear/org.eclipse.jetty.servlet.ServletContextHandler.<init>(ServletContextHandler.java:185) ~[jetty-servlet-11.0.0.jar:11.0.0]
	at deployment.c2oear-snapshot.ear/org.eclipse.jetty.servlet.ServletContextHandler.<init>(ServletContextHandler.java:154) ~[jetty-servlet-11.0.0.jar:11.0.0]
	at deployment.c2oear-snapshot.ear/com.optinuity.c2o.transport.TransportServer.addHandler(TransportServer.java:269) 

At code:
ServletContextHandler cntxHandler = new ServletContextHandler(contextHandlers, context,
				ServletContextHandler.SESSIONS);

Please help to resolve this in Jetty11 itself, since it will be very big effort for me to migrate to Java 17 at this time.

**Comment by sbordet** [comment]
Looks like you are using Jetty embedded, so you need to put the `jakarta.servlet-api.jar` in your class-path.

Double-check your class-path and make sure the jar is there.

**Comment by ilapavuluri** [comment]
@sborde thanks for the quick response, I am using servlet6 whereas Jetty 11.0.20 is on servlet 5, in servlet 6 all deprecated apis are removed, in Jetty 11.0.20 we are still using deprecated api's.

in the above case I am calling " new ServletContextHandler(contextHandlers, context,
ServletContextHandler.SESSIONS);" which intern calls  SessionHandler which uses HttpSessionContext and this is deprecated long back and removed in servlet 6. 

To experiment this instead of calling new ServletContextHandler(contextHandlers, context,
ServletContextHandler.SESSIONS);" I called " new ServletContextHandler(contextHandlers, context);" since this doesn't call SessionHandler, after this the above error is not coming but failing near

ServletHolder sh1 = new ServletHolder(new testclass()); because ServletHolder uses another deprecated api(SingleThreadModel) which is removed in  servlet 6.

To summarize my application runs on servlet 6 where lot of deprecated apis are removed, since Jetty 11 latest still uses those deprecated apis I am facing this issue.

Moving Jetty 12 is big effort for me because Jetty 12 runs only on java 17 & above, and my application runs on Java 11 only.

Appreciate your help on this.

**Comment by gregw** [comment]
@ilapavuluri You want to run jetty-11, which is Servlet 5, but use the Servlet 6 APIs?    This is not a supported configuration as our servlet 6 implementation is jetty-12.

I guess it would not be a huge job to develop a jetty-11.1 release that was jetty-11 compiled against servlet-6.0 with the deprecated APIs removed, but it probably would not be totally spec compliant as there were behaviour clarifications in the spec.

This is not something that we would do on an open source basis.   We might do that for a commercial client (see sales@webtide.com).    Alternately, you are free to give it a go yourself.  I'm guessing it is probably only a few days work (but for us it would then be many more months work releasing documenting supporting.... hence we don't do it unless commercially).

**Comment by ilapavuluri** [comment]
@gregw thank you for your quick response.

**Comment by joakime** [comment]
Note that there are more changes between Servlet 5 and Servlet 6 than just the API.
Example: There are more rules in Servlet 6 with regards to what is allowed in the path portion of the request URI.

See "Rejecting Suspicious Sequences" section in the Servlet 6 spec -> https://github.com/jakartaee/servlet/blob/6.0.0-RELEASE/spec/src/main/asciidoc/servlet-spec-body.adoc
